File: akka-docs/src/test/scala/tutorial_5/DeviceGroup.scala
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package tutorial_5

import akka.actor.{ Actor, ActorLogging, ActorRef, Props, Terminated }
import DeviceGroup._
import DeviceManager.RequestTrackDevice

import scala.concurrent.duration._

object DeviceGroup {
  def props(groupId: String): Props = Props(new DeviceGroup(groupId))

  final case class RequestDeviceList(requestId: Long)
  final case class ReplyDeviceList(requestId: Long, ids: Set[String])

  //#query-protocol
  final case class RequestAllTemperatures(requestId: Long)
  final case class RespondAllTemperatures(requestId: Long, temperatures: Map[String, TemperatureReading])

  sealed trait TemperatureReading
  final case class Temperature(value: Double) extends TemperatureReading
  case object TemperatureNotAvailable extends TemperatureReading
  case object DeviceNotAvailable extends TemperatureReading
  case object DeviceTimedOut extends TemperatureReading
  //#query-protocol
}

//#query-added
class DeviceGroup(groupId: String) extends Actor with ActorLogging {
  var deviceIdToActor = Map.empty[String, ActorRef]
  var actorToDeviceId = Map.empty[ActorRef, String]
  var nextCollectionId = 0L

  override def preStart(): Unit = log.info("DeviceGroup {} started", groupId)

  override def postStop(): Unit = log.info("DeviceGroup {} stopped", groupId)

  override def receive: Receive = {
    //#query-added
    case trackMsg @ RequestTrackDevice(`groupId`, _) ⇒
      deviceIdToActor.get(trackMsg.deviceId) match {
        case Some(ref) ⇒
          ref forward trackMsg
        case None ⇒
          log.info("Creating device actor for {}", trackMsg.deviceId)
          val deviceActor = context.actorOf(Device.props(groupId, trackMsg.deviceId), "device-" + trackMsg.deviceId)
          context.watch(deviceActor)
          deviceActor forward trackMsg
          deviceIdToActor += trackMsg.deviceId -> deviceActor
          actorToDeviceId += deviceActor -> trackMsg.deviceId
      }

    case RequestTrackDevice(groupId, deviceId) ⇒
      log.warning(
        "Ignoring TrackDevice request for {}. This actor is responsible for {}.",
        groupId, this.groupId
      )

    case RequestDeviceList(requestId) ⇒
      sender() ! ReplyDeviceList(requestId, deviceIdToActor.keySet)

    case Terminated(deviceActor) ⇒
      val deviceId = actorToDeviceId(deviceActor)
      log.info("Device actor for {} has been terminated", deviceId)
      actorToDeviceId -= deviceActor
      deviceIdToActor -= deviceId

    //#query-added
    // ... other cases omitted

    case RequestAllTemperatures(requestId) ⇒
      context.actorOf(DeviceGroupQuery.props(
        actorToDeviceId = actorToDeviceId,
        requestId = requestId,
        requester = sender(),
        3.seconds
      ))
  }

}
//#query-added

File: akka-docs-new/src/test/java/jdocs/tutorial_4/DeviceGroup.java
/**
 * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package jdocs.tutorial_4;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.Terminated;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import scala.concurrent.duration.FiniteDuration;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

//#query-added
public class DeviceGroup extends AbstractActor {
    LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

    String groupId;

    public DeviceGroup(String groupId) {
        this.groupId = groupId;
    }

    public static Props props(String groupId) {
        return Props.create(jdocs.tutorial_3.DeviceGroup.class, groupId);
    }

    public static final class RequestDeviceList {
        final Long requestId;

        public RequestDeviceList(Long requestId) {
            this.requestId = requestId;
        }
    }

    public static final class ReplyDeviceList {
        final Long requestId;
        final Set<String> ids;

        public ReplyDeviceList(Long requestId, Set<String> ids) {
            this.requestId = requestId;
            this.ids = ids;
        }
    }

    //#query-protocol
    public static final class RequestAllTemperatures {
        final Long requestId;

        public RequestAllTemperatures(Long requestId) {
            this.requestId = requestId;
        }
    }

    public static final class RespondAllTemperatures {
        final Long requestId;
        final Map<String, TemperatureReading> temperatures;

        public RespondAllTemperatures(Long requestId, Map<String, TemperatureReading> temperatures) {
            this.requestId = requestId;
            this.temperatures = temperatures;
        }
    }

    public static interface TemperatureReading {
    }

    public static final class Temperature implements TemperatureReading {
        public final Double value;

        public Temperature(Double value) {
            this.value = value;
        }
    }

    public static final class TemperatureNotAvailable implements TemperatureReading {
    }

    public static final class DeviceNotAvailable implements TemperatureReading {
    }

    public static final class DeviceTimedOut implements TemperatureReading {
    }
    //#query-protocol


    Map<String, ActorRef> deviceIdToActor = new HashMap<>();
    Map<ActorRef, String> actorToDeviceId = new HashMap<>();
    Long nextCollectionId = 0L;

    @Override
    public void preStart() {
        log.info("DeviceGroup {} started", groupId);
    }

    @Override
    public void postStop() {
        log.info("DeviceGroup {} stopped", groupId);
    }

    @Override
    public Receive createReceive() {
        //#query-added
        return receiveBuilder()
                .match(DeviceManager.RequestTrackDevice.class, trackMsg -> {
                    if (this.groupId.equals(trackMsg.groupId)) {
                        ActorRef ref = deviceIdToActor.get(trackMsg.deviceId);
                        if (ref != null) {
                            ref.forward(trackMsg, getContext());
                        } else {
                            log.info("Creating device actor for {}", trackMsg.deviceId);
                            ActorRef deviceActor = getContext().actorOf(Device.props(groupId, trackMsg.deviceId), "device-" + trackMsg.deviceId);
                            getContext().watch(deviceActor);
                            deviceActor.forward(trackMsg, getContext());
                            actorToDeviceId.put(deviceActor, trackMsg.deviceId);
                            deviceIdToActor.put(trackMsg.deviceId, deviceActor);
                        }
                    } else {
                        log.warning(
                                "Ignoring TrackDevice request for {}. This actor is responsible for {}.",
                                groupId, this.groupId
                        );
                    }
                })
                .match(RequestDeviceList.class, r -> {
                    sender().tell(new ReplyDeviceList(r.requestId, deviceIdToActor.keySet()), getSelf());
                })
                .match(Terminated.class, t -> {
                    ActorRef deviceActor = t.getActor();
                    String deviceId = actorToDeviceId.get(deviceActor);
                    log.info("Device actor for {} has been terminated", deviceId);
                    actorToDeviceId.remove(deviceActor);
                    deviceIdToActor.remove(deviceId);
                })
                //#query-added
                // ... other cases omitted
                .match(RequestAllTemperatures.class, r -> {
                    getContext().actorOf(DeviceGroupQuery.props(
                            actorToDeviceId, r.requestId, getSender(), new FiniteDuration(3, TimeUnit.SECONDS)));
                })
                .build();
    }
}
//#query-added
